### PR TITLE
librados: export stat2 in C API

### PR DESCRIPTION
**Pull request by lightmelodies**
Signed-off-by: Wei Wang <lightmelodies@outlook.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cbodley** [comment]
could you please add a test case for this in `src/test/librados/c_read_operations.cc`? it would look very much like `TEST_F(CReadOpsTest, Stat)`

**Comment by lightmelodies** [comment]
> could you please add a test case for this in `src/test/librados/c_read_operations.cc`? it would look very much like `TEST_F(CReadOpsTest, Stat)`

Done, as well as **libradosstriper**

**Comment by lightmelodies** [comment]
@cbodley any update?

**Comment by cbodley** [comment]
thanks @lightmelodies, this looks good to me but i'll defer to the core team for review and testing - cc @neha-ojha

**Comment by lightmelodies** [comment]
@neha-ojha

**Comment by lightmelodies** [comment]
@tchaikov

File: src/include/rados/librados.h
Comment (tchaikov) [inline src/include/rados/librados.h:2218]: IIUC, the difference between `rados_aio_stat2()` and `rados_aio_stat()` is that the former returns a higher resolution `timespec` while the latter just returns `time_t`. but the new unit tests exercising `rados_aio_stat2()` does not verify this feature? any reason why we don't want to test it?

Comment (lightmelodies) [reply src/include/rados/librados.h:2218]: Add some test to compare `stat` and `stat2`. Tests for `rados_aio_stat2` need additional change, because the current `rados_aio_write_op_operate ` implemention does not set mtime correctly.
https://github.com/ceph/ceph/blob/e70ca62d9cfd5c5891d4327739ba6fc4159b5c03/src/librados/librados_c.cc#L4006

**Comment by lightmelodies** [comment]
@tchaikov any other consideration?

**Comment by tchaikov** [comment]
> @tchaikov any other consideration?

@lightmelodies sorry for the latency. will take a look later today.

File: src/test/librados/c_read_operations.cc
Comment (tchaikov) [inline src/test/librados/c_read_operations.cc:632]: if the following steps rely on the completion of this call, we should use `ASSERT_EQ()` here, like
```c++ 
ASSERT_EQ(0, rados_write_op_operate2(wop, ioctx, obj, &ts, 0));
```

so that the test bails out if `rados_write_op_operate2()` fails.

File: src/test/librados/c_read_operations.cc
Comment (tchaikov) [inline src/test/librados/c_read_operations.cc:638]: ditto.

File: src/test/librados/c_read_operations.cc
Comment (tchaikov) [inline src/test/librados/c_read_operations.cc:643]: drop this empty line.

File: src/test/librados/stat.cc
Comment (tchaikov) [inline src/test/librados/stat.cc:38]: ditto.

File: src/test/libradosstriper/io.cc
Comment (tchaikov) [inline src/test/libradosstriper/io.cc:59]:
```suggestion
  char buf[128] = {};
```

File: src/test/libradosstriper/io.cc
Comment (tchaikov) [inline src/test/libradosstriper/io.cc:56]: i'd suggest always initialize the variable with a "safe" value.

Comment (tchaikov) [reply src/test/libradosstriper/io.cc:56]: the "p" in the variable name could be misleading. might want to use `size` instead.

File: src/test/libradosstriper/io.cc
Comment (tchaikov) [inline src/test/libradosstriper/io.cc:57]: s/pmtime/mtime/.

**Comment by lightmelodies** [comment]
Resolved mostly. Things like ``memset(buf, 0xcc, sizeof(buf));`` are not touched because they are used too widely. Refactoring all of them is far from the intention of this PR.

**Comment by tchaikov** [comment]
jenkins test make check

**Review by tchaikov — APPROVED** [review]

**Comment by sseshasa** [comment]
**Teuthology Test Result**
http://pulpito.front.sepia.ceph.com/?branch=wip-yuri5-testing-2022-08-18-0812
Summary: 294/319 Passed, 22 Failed, 3 dead jobs

**Unrelated Failures**
The following are the unrelated failures for which trackers already exist:

1. https://tracker.ceph.com/issues/57368 - The CustomResourceDefinition "installations.operator.tigera.io" is invalid: metadata.annotations: Too long: must have at most 262144 bytes
2. https://tracker.ceph.com/issues/57122 - test failure: rados:singleton-nomsgr librados_hello_world - Ceph - RADOS
3. https://tracker.ceph.com/issues/55853 - cls_rgw test failures
4. https://tracker.ceph.com/issues/57165 - expected valgrind issues and found none 
5. https://tracker.ceph.com/issues/57270 - stderr E: The repository 'https://download.ceph.com/debian-octopus jammy Release' does not have a Release file
6. https://tracker.ceph.com/issues/52321 - qa/tasks/rook times out: 'check osd count' reached maximum tries (90) after waiting for 900 seconds - Ceph - Orchestrator
7. https://tracker.ceph.com/issues/56850 - src/mon/PaxosService.cc: 193: FAILED ceph_assert(have_pending)

**Comment by yuriw** [comment]
jenkins test this please

**Pull request merged**
